### PR TITLE
State machine local child root cache.

### PR DESCRIPTION
**Pull request by cheme**
This PR adds a local cache for child roots used by state machine.
It is only for 'std' usage. Cumulus will still query trie storage, but it is a in memory trie storage which is fine in this case.

Note that there may be better design or implementing, but this lack of cache is something that did exists for a long time without a good solution, so I am proposing the simpliest approach that comes to my mind.

### CODE REVIEW

**Comment by bkchr** [comment]
Why do we need this? Are there information indicating that this is some sort of bottleneck?

**Comment by cheme** [comment]
No bottleneck indication, just something that is a bit awkward (on each access to child trie we query the root of the child trie without caching so directly on the trie).

**Comment by bkchr** [comment]
> No bottleneck indication, just something that is a bit awkward (on each access to child trie we query the root of the child trie without caching so directly on the trie).

But isn't this the case for any storage access?

**Comment by cheme** [comment]
Storage access uses the cache at client/db/storage_cache.rs.
But this cache only store `key -> value` or `(child_info, key) -> value`
When querying `(child_info, key)`, if value is not in storage_cache.rs we do
- fetch child storage root in top trie
- fetch value in child trie
This PR does cache the first fetch which is redundant for all query with a same child_info.

Ideally the cache for the child trie root would also be done at storage_cache.rs, but the api design makes it really awkward.
(that is something that is known for a long time now, so I choose to write this simple cache instead of having a may never happen TODO).

**Comment by arkpar** [comment]
Looks good but there's a failing test

File: primitives/state-machine/src/trie_backend_essence.rs [hunk 1]
```diff
@@ -80,9 +103,21 @@ impl<S: TrieBackendStorage<H>, H: Hasher> TrieBackendEssence<S, H> where H::Out:
...
 
+	#[cfg(feature = "std")]
+	fn reset_child_roots_cache(&mut self) {
+		self.cache = Arc::new(RwLock::new(Cache::new()));
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn reset_cache(&mut self) {
+		self.cache = Arc::new(RwLock::new(Cache::new()));
+	}
```
Comment (bkchr) [inline primitives/state-machine/src/trie_backend_essence.rs:119]: Both methods are doing the same.

File: primitives/state-machine/src/trie_backend_essence.rs [hunk 2]
```diff
@@ -46,11 +50,28 @@ pub trait Storage<H: Hasher>: Send + Sync {
 	fn get(&self, key: &H::Out, prefix: Prefix) -> Result<Option<DBValue>>;
 }
 
+#[cfg(feature = "std")]
+/// Local cache for child root.
```
Comment (bkchr) [inline primitives/state-machine/src/trie_backend_essence.rs:54]:
```suggestion
/// Local cache for child root.
#[cfg(feature = "std")]
```

**Review by bkchr — APPROVED** [review]

File: primitives/state-machine/src/trie_backend_essence.rs [hunk 3]
```diff
@@ -80,9 +103,21 @@ impl<S: TrieBackendStorage<H>, H: Hasher> TrieBackendEssence<S, H> where H::Out:
...
 	}
 
+	#[cfg(feature = "std")]
+	fn reset_child_roots_cache(&mut self) {
+		self.cache = Arc::new(RwLock::new(Cache::new()));
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn reset_cache(&mut self) {
+		self.cache = Arc::new(RwLock::new(Cache::new()));
```
Comment (cheme) [inline primitives/state-machine/src/trie_backend_essence.rs:118]:
```suggestion
```

File: primitives/state-machine/src/trie_backend_essence.rs (hunk 1 above)
Comment (cheme) [reply primitives/state-machine/src/trie_backend_essence.rs]: :+1: , explains CI failures.

File: primitives/state-machine/src/trie_backend_essence.rs [hunk 4]
```diff
@@ -80,9 +103,21 @@ impl<S: TrieBackendStorage<H>, H: Hasher> TrieBackendEssence<S, H> where H::Out:
...
 		self.root = root;
 	}
 
+	#[cfg(feature = "std")]
+	fn reset_child_roots_cache(&mut self) {
+		self.cache = Arc::new(RwLock::new(Cache::new()));
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn reset_cache(&mut self) {
```
Comment (cheme) [inline primitives/state-machine/src/trie_backend_essence.rs:117]:
```suggestion
	fn reset_child_roots_cache(&mut self) {
```

File: primitives/state-machine/src/trie_backend_essence.rs [hunk 5]
```diff
@@ -80,9 +103,20 @@ impl<S: TrieBackendStorage<H>, H: Hasher> TrieBackendEssence<S, H> where H::Out:
...
 	}
 
+	#[cfg(feature = "std")]
+	fn reset_child_roots_cache(&mut self) {
+		self.cache = Arc::new(RwLock::new(Cache::new()));
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn reset_child_roots_cache(&mut self) {
+	}
```
Comment (bkchr) [inline primitives/state-machine/src/trie_backend_essence.rs:118]:
```suggestion
	fn reset_child_roots_cache(&mut self) {}
```

File: primitives/state-machine/src/trie_backend_essence.rs [hunk 6]
```diff
@@ -80,9 +103,20 @@ impl<S: TrieBackendStorage<H>, H: Hasher> TrieBackendEssence<S, H> where H::Out:
...
 	/// Set trie root. This is useful for testing.
 	pub fn set_root(&mut self, root: H::Out) {
+		// if root did change so can have child roots.
+		self.reset_child_roots_cache();
 		self.root = root;
 	}
 
+	#[cfg(feature = "std")]
+	fn reset_child_roots_cache(&mut self) {
+		self.cache = Arc::new(RwLock::new(Cache::new()));
```
Comment (bkchr) [inline primitives/state-machine/src/trie_backend_essence.rs:113]: This also doesn't reflect the name of the function. If this should only reset the child roots, you should do it. Currently you clear the entire cache.

File: primitives/state-machine/src/trie_backend_essence.rs [hunk 7]
```diff
@@ -79,9 +104,19 @@ where
 
 	/// Set trie root. This is useful for testing.
 	pub fn set_root(&mut self, root: H::Out) {
+		// If root did change so can have cached content.
+		self.reset_cache();
```
Comment (bkchr) [inline primitives/state-machine/src/trie_backend_essence.rs:108]: Don't we also need to reset the cache after having run `storage_root`?

Comment (cheme) [reply primitives/state-machine/src/trie_backend_essence.rs:108]: Should be fine, the cache is above the trie backend so bellow all overlay and ext stuff, it really only change when trie change (after client does process block when state machine instance is already dropped).
Here it is related to the way it was use in some test function (changing root means basically changing the trie backend).

**Comment by bkchr** [comment]
bot merge

**Pull request merged**